### ISSUE TITLE
Built-in support of alert rules depending on other alert rules

### ISSUE THREAD

**Issue by jaimegago** [grafana/grafana#10901]
**Use case**
As a Grafana user I'd like to define an alert which state ("ok" or "alerting") is calculated from states of existing alerts and that is evaluated automatically every N seconds.

**Implementation proposal**
Create a new panel type (that will need to tap into the alerting engine)

Inputs
- select existing alerts
- select either of "AND/OR" operators between the selected alerts
- associate existing notification channels with the so called "alert group"
- define a "refresh" duration

Output
- Displays the current state of the alert group
- Displays each alert in the group in their current state with links to them
- Displays the configured operator between each alert
- Displays the refresh duration
- Notifies the specified notifications channels on state changes (the same as a regular alert does)

A history of the Alert Group state and a "test" button in the same fashion as the alert tab of a panel would be nice to have.

**Comment by torkelo** [comment]
I have been thinking about this since the we implemented the initial alerting feature. The way I hoped to implement it was to add a new alert condition type (other than the current Query type), that can check another alert rule.

**Comment by jaimegago** [comment]
This is a feature that would be definitely tightly coupled to the alerting engine so I understand trying to keep them together as much as possible. With your implementation, what would the panel look like when I create a new alert that is only made of this new type?

**Comment by torkelo** [comment]
Well there are two uses cases, combine alert rule state conditions with your metric query condition, would just be a new condition.

If you want to create an alert rule out of only other alert rules , then I am not sure. Maybe a graph showing alert rule state history :)

**Comment by jaimegago** [comment]
> Well there are two uses cases, combine alert rule state conditions with your metric query condition, would just be a new condition.
If there was a dedicated panel for alerting based on alert state then I could still replicate the mix and matching use case by creating an alert based on a regular query (or queries) and then use that alert state in my "alert group" panel. My point being though my proposed implementation is different it covers the same use cases.

>If you want to create an alert rule out of only other alert rules , then I am not sure. Maybe a graph showing alert rule state history :)
So you would have a dedicated code path in the current alert UI for when the alert rule is composed exclusively of checking other alerts ?

**Comment by torkelo** [comment]
> If there was a dedicated panel for alerting based on alerts state then I could still replicate the mix and matching 

But that feels very cumbersome! If wanted to create an alert rule (A) on say application errors, but I only want it to fire if another alert rule (B, for database access)  was green. Then I would have to create 3 rules, one for A, one for B, and a C that combined them. 

I think creating a new panel for this should not be the first step. Simply creating a new alert condition, something the existing alerting engine was designed for should be the first step. Then see if there is demand / use cases for a separate panels to only combine alert rules.

**Comment by jaimegago** [comment]
Yes with the new panel it would take an extra step/rule/panel, now while I think we can argue about the cumbersomeness of each way of doing it, I don't see a right or wrong answer here. But I'm willing to continue lining up pros and cons of each if you are =D.
In your example where would you put the condition on the other alert on A or on B, why?  You could say it doesn't matter as they result in the same but from the perspective of a guy that gets paged for a rule he didn't create it'd be clearer with a C alert that combines them. Now nothing prevents me with your implementation to create that 3rd C rule and just fill it with an alert of alerts and I'll advise "my" users to do just that, i.e. don't combine types in your panels, just because you can doesn't mean that you should.
Another problem with your example is what if I want A and B to continue to be evaluated by themselves? I would then still need to create my C rule that links them. That scenario is the one that got me to fill this issue in the first place, and probably why I thought about a new panel. Imagine a team that has existing alerts they want to group without changing them and while trying to be as clear as possible when it comes to being paged.

**Comment by Siecje** [comment]
Is it possible to have an alert send a notification every hour or once a day if the alert is still alerting?

**Comment by jaimegago** [comment]
@Siecje Github issues are not the right place for your question, in the future please use [the community website](https://community.grafana.com), you'll get better support.
The answer to your question is: not at the moment.

**Comment by torkelo** [comment]
It’s not matter of pros or cons, the alert rules are composed of conditions that are evaluated. So even for a new panel type you would need a new condition type. One that could be used in normal graph alert rules. That is why I want to start with that and see how that can be used and then review demand / need for another panel

**Comment by jaimegago** [comment]
I am not saying there are pros and cons to the new condition type, it's how you choose to expose that new condition type that in my not so humble opinion has pros and cons.
Again though having that new condition available regardless of the UI implementation (new panel type or not) would be awesome and completely fulfill the requirements.
I'm just trying to provide as much context and user feedback (I've been talking to Grafana alerts users) before hand in the hopes that it helps as anecdotal as it can be =).

**Comment by bergquist** [comment]
I think this is the wrong way to go. Alert rules should be easy to reason about, test and verify. Alert rules depending on other alert rules seems to increase the complexity quite a lot. 

Before we consider this I would like to have multiple concrete examples where this is actually needed.

**Comment by nilium** [comment]
I don't know that this is still being considered, but there's at least one example scenario I'd like to prevent with being able to have one alert rule depend on another.

In the event that a primary service goes down, that service will trigger an alert in Grafana. Because this service goes down, several other dependent services will begin triggering an alert because that one service is down -- after all, they depend on it.

If all these alerts go off at roughly the same time, then it becomes difficult to make sense of the noise and interrupts any work being done to fix the primary service. What I think I'd like to have happen is only the service that needs to send an alert does so, and something like this proposal might allow that.

By making the dependent services' alerts depend on the primary service _not_ alerting, we'd be able to limit the amount of noise that an engineer has to filter during a 3am incident.

The counter-argument to this is that everyone ought to just engineer their systems better so that every alert is very specific. Another way to avoid this is to simply create an invisible query in each graph that alerts and invert the alerting rules of the dependent alert, but this is error-prone (what if someone doesn't update one of them when the primary alert changes?) and probably not what anyone wants to do.

**Comment by bmcgough** [comment]
I am interesting in this as a feature as well, and have another use case.

We have a system where the loss of redundancy for a period of time is an acceptable risk. In that state (redundancy lost, Grafana Alerting) however, there are other alerts that would change the risk calculation. In short, alert a is not enough to send notification, but alert a and alert b (or c, or d) may be enough to notify.

How about implementing alerts in the Alert List panel? It doesn't have queries defined, but seems to be able to harvest alerts across the dashboard. That seems like a clean place to do it, and makes it pretty clear how the logic all works.

**Comment by theodiefenthal** [comment]
Here is another usecase: 
I have a processing pipeline with multiple stages. I have e.g. ten "input" kafka topics, each with 20 partitions. Job 1 reads data from those topics, parses and processes it and then publishes the data to "processed" topics. 
Following are multiple downstream jobs, each reading from multiple "processed" topics, combine the data and do their work. 

Best monitoring we can do is to monitor the kafka cluster / consumers directly. If the consumers advance properly, we know for sure that the system is working and all is fine. If they don't advance and build up lag (or there are no messages coming in), something is broken and we need to alert. 

Now I have multiple scenarios benefitting from this feature here:
Scenario 1: If I'm missing data from one of my input kafka topics in ONE partition, pretty sure there is a server down delievering the data => Alert. 
If there is no data for the entire topic, i.e. for all 20 partitions, there is a higher level problem (e.g. firewall rule or certificates expired, data center down or similar). I then don't need an alert for each of the 20 partitions in that topic but a higher severity one for the entire kafka topic. As others already wrote, producing 20 alerts in that case will just create a lot of noise. 

Scenario 2: If there is no data coming in for one input topic, all the downstream jobs which rely on that data can't progress as well. So I don't want to generate alarms for all those jobs if I know that job 1 is already in alerting state. 

I think, having such a feature would be beneficial to all streaming pipelines and event driven systems.

**Comment by armandgrillet** [comment]
This will require sequential evaluation for alert rules in a group which is planned.
